iwyu_output.cc

```diff
@@ -1105,13 +1105,28 @@ void ProcessForwardDeclare(OneUse* use,
     }
   }
 
-  // (A7) If --no_fwd_decls has been passed, recategorize as a full use unless
-  // the decl is in this file (in which case it must be a self-sufficient decl
-  // being used, so we can just let IWYU do its work).
-  if (!use->ignore_use() &&
-      GlobalFlags().no_fwd_decls &&
-      GetFileEntry(use->decl_loc()) != GetFileEntry(use->use_loc())) {
-    use->set_full_use();
+  // (A7) If --no_fwd_decls has been passed, and a decl can be found in one of
+  // the headers, suggest that header, and recategorize as a full use. If we can
+  // only find a decl in this file, it must be a self-sufficent decl being used,
+  // so we can just let IWYU do its work, and there is no need to recategorize.
+  if (!use->ignore_use() && GlobalFlags().no_fwd_decls) {
+    bool promote_to_full_use = true;
+    for (const Decl* decl = use->decl(); decl != nullptr;
+         decl = decl->getPreviousDecl()) {
+      if (IsBeforeInSameFile(decl->getLocation(), use->use_loc())) {
+        promote_to_full_use = false;
+      } else if (IsBeforeInTranslationUnit(decl->getLocation(),
+                                           use->use_loc())) {
+        use->reset_decl(cast<NamedDecl>(decl));
+        use->set_suggested_header(ConvertToQuotedInclude(GetFilePath(decl)));
```

Comment:
I suspect this would need to be run through the mapping machinery again -- for example, if symbol `X` is declared in private header `x_internal.h`, and documented to be used from `x.h`, we'll rewrite whatever the earlier analysis came up with to `"x_internal.h"`, because that's where the decl is.

I'm not sure how/when/where mapping happens now (I really should), so I don't know if there's a simple way to just remap to the public header. If you don't find an obvious way, I think it would be fine to leave a `TODO` here.

Reply:
I think this can be solved by not setting a suggested header, so the header will be chosen in `CalculateMinimalIncludes`.

However there is another problem (they are maybe related, I'm not sure):
```
// foo.h
class Foo;

// fooandbar.h
class Foo;
class Bar;

// test.cpp
#include "fooandbar.h"
#include "foo.h"

Foo *foo;
Bar *bar;
```
Here IWYU will say that the includes are correct, but in fact `foo.h` could be removed. (If you change the order of the includes IWYU will correctly suggest the removal.)
I left a `TODO` for this one.

Reply:
Interesting. This keeps coming up (see the tolerate-transitive PR), where decisions made at one point are invalidated by later steps. But I think a TODO is fine for this. If we manually remove `foo.h`, IWYU won't complain, so that's cool.